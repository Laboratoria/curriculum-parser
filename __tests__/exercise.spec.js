const path = require('path');
const helpers = require('./helpers');
const exercise = require('../lib/exercise');


describe('exercise', () => {
  it('should reject when missing readme', () => {
    expect.assertions(1);
    return exercise(helpers.resolveFixturePath('00-part-missing-readme'))
      .catch((err) => {
        expect(err.message).toBe('Empty or missing exercise README.md');
      });
  });

  it('should parse exercise', () => (
    exercise(
      path.join(
        helpers.resolveFixturePath('paradigms'),
        '01-paradigms',
        '04-imperative-exercise',
        '01-print-primes',
      ),
    )
      .then((result) => {
        expect(result).toMatchSnapshot();
      })
  ));
});
