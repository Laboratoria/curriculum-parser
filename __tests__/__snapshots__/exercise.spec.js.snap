// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exercise should parse exercise 1`] = `
Object {
  "body": "<p>La programación imperativa tiene un nivel de abstracción bajo, lo que quiere
decir que se enfoca en describir, desde el punto de vista de la computadora,
cómo hacer lo que queremos. Nuestro código expresa de forma explícita el control
de flujo, usando bucles, declara variables para llevar la cuenta, ...</p>
<p>En el mundo real, en un lenguaje como JavaScript, vamos a encontrar
características de este paradigma, pero siempre en el contexto de otro paradigma
con un nivel de abtracción más alto y un mecanismo de reuso de código (cómo la
programación orientada a objetos o la prgramación funcional).</p>
<h2 id=\\"tarea\\">Tarea</h2>
<p>El <em>boilerplate</em> (la plantilla sobre la que hacer el ejercicio) contiene un
programa sin terminar. Este programa debe <strong>imprimir en la consola todos los
números primos menores que <code>20</code></strong>.</p>
<p>El <em>boilerplate</em> impementa sólo parte de la lógica. Por el momento imprime todos
los números enteros del <code>2</code> al <code>19</code>. Ahora tienes que añadir la lógica para
comprobar si un número es primo o no, y así sólo imprimir los primos.</p>
<h2 id=\\"requisitos\\">Requisitos</h2>
<ul>
<li>Debes implementar un bucle <code>for</code> anidado dentro del <code>for</code> ya existente.</li>
</ul>
<h2 id=\\"pistas\\">Pistas</h2>
<ul>
<li>Si un número es divisible por cualquier entero mayor que <code>1</code> que no sea el
número mismo, sabemos que NO es primo.</li>
<li>Dentro del bucle <code>for</code> principal, para ver si un número es primo, vamos a
tratar de demostrar que NO es primo, y si no podemos probar que no es primo,
entonces podemos asumir que sí es primo.</li>
<li>El operador <code>%</code> nos da el residuo de una división.</li>
<li>Una vez que hemos probado que un número no es primo ya no necesitamos
continuar dividiéndolo.</li>
</ul>
<h2 id=\\"recursos\\">Recursos</h2>
<ul>
<li><a href=\\"https://goo.gl/zn7cTu\\" target=\\"_blank\\">Arithmetic Operators: Remainder - MDN</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for\\" target=\\"_blank\\">Statements: for - MDN</a></li>
</ul>
",
  "env": "js",
  "files": Object {
    "Ym9pbGVycGxhdGU=": Object {
      "cHJpbnRQcmltZXMuanM=": "function printPrimes() {
  // Iteramos desde 2 (el primer número primo) hasta 19 (menor que 20)
  for (var i = 2; i < 20; i++) {
    var isPrime = true;

    // Ahora debemos iterar desde \`2\` hasta \`i / 2\` usando una variable \`j\` como
    // contador.
    //
    // Para cada \`j\` comprueba si la división \`i / j\` deja un residuo.
    //
    // NOTA: No necesitamos probar números más allá de \`i / 2\` porque el
    // resultado de la división \`i / j\` siempre sería una fracción entre 2 y 1,
    // siempre dejando residuo.

    // TU CÓDIGO AQUÍ

    if (isPrime) {
      console.log(i);
    }
  }
}

module.exports = printPrimes;
",
    },
    "c29sdXRpb24=": Object {
      "cHJpbnRQcmltZXMuanM=": "function printPrimes() {
  for (var i = 2; i < 20; i++) {
    var isPrime = true;
    for (var j = 2; j <= (i / 2); j++) {
      if (i % j === 0) {
        isPrime = false;
        break;
      }
    }
    if (isPrime) {
      console.log(i);
    }
  }
}

module.exports = printPrimes;
",
    },
    "dGVzdA==": Object {
      "cHJpbnRQcmltZXMuc3BlYy5qcw==": "'use strict';


const Assert = require('chai').assert;
const Sinon = require('sinon');
const Submission = require('../solution/printPrimes');


describe('printPrimes()', () => {

  const stub = Sinon.stub(console, 'log');
  Submission();
  stub.restore();

  it('debería exportar una función', () => {
    Assert.equal(typeof Submission, 'function');
  });

  it('debería invocar console.log 8 veces', () => {
    Assert.equal(stub.callCount, 8);
  });

  [[2], [3], [5], [7], [11], [13], [17], [19]].forEach((args, idx) => {
    it(\`invocación \${idx + 1} a console.log debería recibir \\"\${args[0]}\\" como único argumento\`, () => {
      Assert.deepEqual(args, stub.getCall(idx).args);
    });
  });

});
",
    },
  },
  "slug": "01-print-primes",
  "title": "Imprime los número enteros menores que 20",
}
`;
